Update SEP-0024 claimable balance op support (#757)
stellar/stellar-protocol: ecosystem/sep-0024.md

`wallet_name` | string | (optional) In communications / pages about the deposit, anchor should display the wallet name to the user to explain where funds are going.
`wallet_url` | string | (optional) Anchor should link to this when notifying the user that the transaction has completed.
`lang` | string | (optional) Defaults to `en`. Language code specified using [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1). `error` fields in the response, as well as the interactive flow UI and any other user-facing strings returned for this transaction should be in this language.

> `claimable_balance_supported` | boolean | (optional) True if submitting transactions as a claimable balance is accepted, false otherwise.

Reviewer: >`claimable_balance_supported`...a claimable balance is accepted...

Does this field indicate to the server that the client supports claimable balances?

Author: That is correct.

[Claimable Balance](https://www.stellar.org/blog/protocol-14-improvements) is an optional feature for both wallets and anchors which enable payments to accounts that don't trust the asset being sent.
This feature splits a payment into two separate parts: the creation of a balance, and the claiming of a balance.
A balance can be claimed asynchronously by the designated claimant (user) when they're ready.
This feature provides less friction for Stellar accounts who wish to deposit external assets with an anchor without making a trustline to the anchor's tokenized asset beforehand.
Anchors will no longer have to wait until the user's Stellar account has trustline to the sending asset before sending funds. Instead, anchors can make the payment using a Claimable Balance operation and the user's Stellar account can claim the funds at their own convenience.

**NOTE** in the *future* the following claimable balance implementations will be mandatory in order to be SEP-0024 interoperable.

**Wallets**: To support claimable balances wallets must

* Send the additional `claimable_balance_supported` request parameter in the POST /transactions/deposit/interactive request body.

```json
{
  ...

>   "claimable_balance": true

Reviewer 2: Shouldn't this be `claimable_balance_supported` ?

}
```

* Periodically poll for account's available claimable balances.
* Provide a UI that allows users to claim a claimable balance.

**Anchors**: To support claimable balances anchors must

* Accept the `claimable_balance_supported` request parameter in `POST /transactions/deposit/interactive` requests and submit deposit transactions using Claimable Balance operations to Stellar accounts that don't yet have a trustline to the asset requested.
* Add `claimable_balance_id` to their [deposit transactions responses.](####Fields-for-deposit-transactions)

**Anchors and Wallets**: Both anchors and wallets must also support the aforementioned ["Stellar account doesn't have a trustline for that asset."](#stellar-account-doesnt-trust-asset) flow.
Wallets need to be interoperable with anchors that have not adopted the claimable balance feature and vice versa.

##### Wallet Claimable Balance Flow

Wallets follow the [typical deposit flow](#Basic-Wallet-Implementation) but will provide the `claimable_balance_supported` request parameter in their POST parameters for `/transactions/deposit/interactive`.

```json
{
  ...

>   "claimable_balance": true

Reviewer 2: Shouldn't this be `claimable_balance_supported` ?

3. User sends external assets to the anchor's off-chain account.
4. Anchors detect that the user's Stellar account doesn't have a trustline.
5. From there the anchor will continue the [deposit flow](#basic-anchor-implementation), crafting the transaction as a claimable balance operation.

##### Anchor Feature: Extend Transaction history endpoints For Claimable Balances

Wallets query the [transaction endpoint](#transaction-history) to see and display the status of deposits.
Anchors should extend their transaction(s) endpoint response to include the `claimable_balance_id` that the anchor added to the [transaction field.](####Fields-for-deposit-transactions)
This will help wallets detect that the user's deposit was successful and theres an [available claimable balance for them.](#wallet-feature-claimable-balance-polling)

##### Anchor Claimable Balance Claimants and Predicates

Predicates are one of the [claimable balance parameters](https://developers.stellar.org/docs/glossary/claimable-balance/#parameters) used to craft a Claimable Balance transaction. They're conditions to meet in order for the recipient to claim their balance.

> It's recommended to use the `UNCONDITIONAL` predicate to allow the claimant to claim the balance without any barriers.

Reviewer 2: Why is putting some sane predicates (collect within 30 days, for example) not recommended?

Author: If the account doesn't claim the balance in 30 days the balance unfortunately doesn't refund back into the sender's account. It's lost for both parties.
In order to provide a friendlier experience for the claimant/receiver its best not to restrict them from when they want to establish their trustline and claim.

If the claimant cannot claim the balance due to time constraints or whatnot then we would have to recommend the following for Wallets and Anchors

- Wallets need to provide more aggressive messaging to their users that there's a deadline to claim their balance
- Wallets and Anchors need to provide ways to refund deposits in the scenario where a claimablebalance is dead on the DEX
- (or) Anchors need refund themselves by issuing more of their asset on the network to send another claimable balance to the accounts that forget to claim in 30 days

Reviewer 3: I think the simplest option would to have the sponsor list themselves as a claimant, and claim the balance if it's outstanding for too long.

Author: I think thats an interesting idea.
Maybe we should see what the community thinks over time.

Reviewer 2: I prefer not having a recommendation rather than recommending something that we're not sure is the best practice

Reviewer 2: we can leave it up for the anchor for now

Author: @tomerweller  what do you think of this?
```suggestion
Anchors are free to set whichever predicates they feel are necessary in order to claim the balance.
If there are no predicate preferences, `UNCONDITIONAL` allows accounts to claim balances at anytime.
```

Reviewer 2: LGTM!

The a new ledger entry splits a payment into two separate parts: the creation of a balance, and the claiming of a balance.
A balance can be claimed asynchronously by the designated claimant (user) when they're ready.
This feature provides less friction for Stellar accounts who wish to deposit external assets with an anchor without making a trustline to the anchor's tokenized asset beforehand.
Anchors will no longer have to wait until the user's Stellar account has trustline to the sending asset before sending funds. Instead, anchors can make the payment using a ClaimableBalance operation and the user's Stellar account can claim the funds at their own convenience.

**NOTE** in the *future* the following claimable balance implementations will be mandatory in order to be SEP-0024 interoperable.

**Wallets**: To support claimable balances wallets must

* Send `claimable_balance_supported` request parameter in their POST /transactions/deposit/interactive request.

```json
{
"claimable_balance": true
}

> ```

Author: `commit suggestions` is sus sometimes haha

2. When the transaction status becomes `pending_user_transfer_start` send the required payment as described in the interactive webapp callback or the `/transaction` endpoint.
3. The anchor will ignore the `claimable_balance_supported` field, discover the account does not have a trustline, and report transaction status as `pending_trust`.
4. At this point both the wallet and anchor are waiting on the trustline to be established in order to continue the deposit payment. This is described here [steps 4-6](#stellar-account-doesnt-trust-asset)

##### Wallet Feature: Claimable Balance Polling

Polling allows the wallet to detect claimable balances and provide a UI to users that there is a balance available to claim. See the ["Claim Account Claimable Balance"](#wallet-feature-claim-account-claimable-balance) section to learn more about how to claim a balance.
There are a few options as to how a wallet can poll and surface claimable balance information for their users.
1. The wallet can poll the anchor's [transaction history endpoints](#anchor-feature-extend-transaction-history-endpoints-for-claimable-balances)

> 

Reviewer 3: ```suggestion
```

2. The wallet can poll Horizon's [`/claimable_balances/`](https://developers.stellar.org/api/resources/claimablebalances/list/) endpoint (or Horizon's [`/claimable_balances/`](https://developers.stellar.org/api/resources/claimablebalances/single/) single endpoint) to pull any available claimable balances for the claimant.

> 

Reviewer 3: ```suggestion
```

   * The single endpoint needs the `claimable_balance_id` in order to retrieve the claimable balance info.
     * This can be obtained by extracting the `claimable_balance_id` from `results_xdr` of the deposit transaction submitted if the anchor does not extend their [transaction history endpoints](#anchor-feature-extend-transaction-history-endpoints-for-claimable-balances)

##### Wallet Feature: Claim Account Claimable Balance

> It's recommended that Wallets provides a UI to enable accounts to claim their balances [polled](#wallet-feature-claimable-balance-polling)
> 
> [Example on how to claim a balance](https://developers.stellar.org/docs/glossary/claimable-balance/#claim-claimable-balance)
> Accounts will still need to establish trustline to the asset before claiming the outstanding balance.

Reviewer 3: Make sure you remove this code if you commit your suggestion:
```suggestion
```

Author: Remove only this correct?
```
[Example on how to claim a balance](https://developers.stellar.org/docs/glossary/claimable-balance/#claim-claimable-balance)
```

Reviewer 3: No, I think you need to introduce the "the account needs to have a trustline prior to claiming a balance" before you show the code.

##### Anchor Claimable Balance Flow

1. Wallets make a request to `/transactions/deposit/interactive` and provide the `claimable_balance_supported` request parameter.

> 2. Anchors mark the transaction as [`claimable_balance_supported=true`](#deposit)

Reviewer 3: I think this needs to be more explicit.
```suggestion
2. Anchors update their internal database record of the transaction to indicate the wallet supports receiving a claimable balance
```

> 3. Anchors detect that the user's Stellar account doesn't have a trustline.

Reviewer 3: Can we add a step before this one like `User sends funds to the anchor's off-chain account`?

> 4. The anchor should then report transaction status as `pending_anchor` and submit a Stellar transaction containing a create claimable balance operation

Reviewer 3: I think we can remove this line because:
1. The transaction status should already be in `pending_anchor` while the anchor checks for a trustline, and
2. The correct status once the anchor begins submitting the transaction is `pending_stellar`.

5. From there the anchor will continue the [deposit flow](#basic-anchor-implementation)

##### Anchor Feature: Extend Transaction history endpoints For Claimable Balances

Wallets query the [transaction endpoint](#transaction-history) to see and display the status of deposits.
Anchors should extend their transaction(s) endpoint response to include the `claimable_balance_id` that the anchor added to the [transaction field](####Fields-for-deposit-transactions).
This will help wallets to provide UI to showcase that their user's deposit was successful and theres an [available claimable balance for them.](#wallet-feature-claimable-balance-polling)

```
GET TRANSFER_SERVER_SEP0024/transactions

> ```

Reviewer 3: I don't think these examples are necessary, we already give json examples in the transaction section. Just add `null` and the non-null value to those examples. Anchors who don't implement claimable balances will either not add the attribute or keep it `null`.

      "external_transaction_id": "2dd16cb409513026fbe7defc0c6f826c2d2c65c3da993f747d09bf7dafd31093",
      "claimable_balance_id": "00000000c2d8c89264288dbde8488364fd3fd30850fd4e7fbf6d1e9809702558afa4fdea",
    }
  ]
}
```

```
GET TRANSFER_SERVER_SEP0024/transaction
```

Example response:

```json
{
  "transaction": {
      "id": "82fhs729f63dh0v4",
      "kind": "deposit",
      "status": "completed",
      "status_eta": 3600,
      "external_transaction_id": "2dd16cb409513026fbe7defc0c6f826c2d2c65c3da993f747d09bf7dafd31093",
      "more_info_url": "https://youranchor.com/tx/242523523",
      "amount_in": "18.34",
      "amount_out": "18.24",
      "amount_fee": "0.1",
      "started_at": "2017-03-20T17:05:32Z",
      "claimable_balance_id": "00000000c2d8c89264288dbde8488364fd3fd30850fd4e7fbf6d1e9809702558afa4fdea",
    }
}
```

##### Anchor Claimable Balance Claimants and Predicates

 Predicates are one of the [Claimable balance parameters](https://developers.stellar.org/docs/glossary/claimable-balance/#parameters) used when craftting a Claimable Balance transaction.
They're used as conditions to meet in order for the recipient to claim their balance.

>       It's recommended to use the `UNCONDITIONAL` predicate to allow the claimant to claim the balance without any barriers.

Reviewer 3: Remove the indent. I've noticed you like putting 1 sentence per line too, which is fine I guess but it may not be necessary. There's nothing wrong with multi-sentence lines.

5. `Anchor` detects the trust line creation in the `Account`. If the asset is `AUTH_REQUIRED`, `Anchor` approves the new trust line.
6. `Anchor` proceeds with the deposit flow.

> ##### Claimable Balances

Reviewer 3: There are a lot of weird line breaks in this section

> [Protocol
> 14/15](https://www.stellar.org/blog/protocol-14-improvements?locale=en) now allows anchors to
> complete payments using the [ClaimableBalance
> functionality](../core/cap-0023.md) and the user can collect the funds at any
> time.
> 
> This is an optional, yet recommended, feature that will no longer need to wait
> for the `Wallet` or the user to establish a
> trust line for the `Account` for the given asset.

Reviewer 3: ```suggestion
[Claimable Balances](https://www.stellar.org/blog/protocol-14-improvements) allow for payments to accounts that do not have a trustline for the asset sent. 

Anchors can optionally use this functionality to send SEP-24 deposit transactions to Stellar accounts that don't yet have a trustline to the asset they have requested.

Wallets should poll for their users' received Claimable Balances so that users can "claim" them. Claiming a Claimable Balance adds a trustline to the account for the asset claimed and adds the amount transacted.
```

> 1. If the given Stellar `account` does not exist, on receipt of the deposit, the

Reviewer 3: I don't think these 2 points are necessary. The logic is simple, no trustline --> claimable balance, trustline --> payment. I wouldn't introduce the "if the stellar account doesn't exist" case.

5. `Anchor` detects the trust line creation in the `Account`. If the asset is `AUTH_REQUIRED`, `Anchor` approves the new trust line.
6. `Anchor` proceeds with the deposit flow.

##### Claimable Balances

[Claimable Balances](https://www.stellar.org/blog/protocol-14-improvements) allows payments to accounts that do not have a trust-line for the asset sent.

**Wallets** poll for their users' received Claimable Balances so that users can "claim" them.

> **Anchors** can optionally use this functionality to send SEP-24 deposit transactions to Stellar accounts that don't yet have a trust-line to the asset they have requested.

Reviewer 3: I think we should emphasize that supporting claimable balances is currently _optional_ for anchors but _mandatory_ for wallets. Wallets reading this need to infer that some of their partner anchors may implement claimable balances, and therefore need to support it as well, but it's not super clear.

**Wallet Claimable Balance Flow**

Wallets will still follow the [typical deposit flow](#Basic-Wallet-Implementation).
The POST request however will extend to include a `claimable_balance` parameter

```
POST TRANSFER_SERVER_SEP0024/transactions/deposit/interactive
Content-Type: multipart/form-data
```

New Request Parameter:

Name | Type | Description
-----|------|------------

> `claimable_balance` | boolean | True if submitting transactions as a claimable balance is accepted, false otherwise.

Reviewer 2: `claimable_balance` is not really self descriptive - it doesn't convey that cap23 is supported, not required. while a bit cumbersome I would go with something like 
`is_calimable_balance_supported` or `is_claimble_balance_accepted`. You can drop the `is` if it's over the top and go with something like `claimable_balance_accepted`.

**Wallet Claimable Balance Polling**

> The wallet must either listen for a callback from the anchor with the

Author: I was suggesting that anchors append `claimable_balance_id` to the deposit response.

> The wallet must either listen for a callback from the anchor with the

Reviewer 3: Oh, we have several real callbacks in this SEP so lets not use that word.

`claimable_balance_id` or poll Horizon's
[`/claimable_balances/`](https://developers.stellar.org/api/resources/claimablebalances/list/)
endpoint to pull any available claimable balances for the claimant.

*Recommended request*

> `GET /claimable_balances?claimant={:claimant}&order={asc,desc}&limit={1-200}`

Author: > Does this request change depending on the architecture of the client's system? 

I don't anticipate that the GET request to horizon would change. This is an example of how the client can discover that there's CB available for the account.

>Could a client extract the claimable balance ID from the `results_xdr` and just use it in the GET request?

Either way is possible.
The GET request surfaces _all_ of the account's available CB; from there they could communicate what those id(s) are to the account holder.
or
They can decode the `results_xdr` of the transaction submitted, discover that it was submitted as a claimable balance, then pull the balance id from the xdr.

> `GET /claimable_balances?claimant={:claimant}&order={asc,desc}&limit={1-200}`

Author: > I wouldn't try to anticipate how the client will want to poll, but I think outlining the 1-3 common approaches would be helpful.

I think thats a great idea.

This allows the wallet to update the claimant that their is a balance available

> to claim and they may need to establish a [trust-line](#stellar-account-doesnt-trust-asset) in order to do so.

Author: Apologies my editor tries to keep line length under 80.
I extended it so it shouldn't be a problem in the future.

**Anchor Claimable Balance Flow**

> 1. The given Stellar `account` does not have a trust-line to the asset's issuer account.

Author: That's correct.
The sender can make as many payments (of an asset that the recipient doesn't trusts) that they want as long as their CBs.
The recipient has to trust the asset in order to add that balance to their account.

2. The `Anchor` proceeds with the deposit flow via the [CAP-0023](../core/cap-0023.md) Create Claimable Balance operation.

**Anchor Claimable Balance Field Changes**

### Fields for deposit transactions

Name | Type | Description
-----|------|------------

> `claimable_balance_id` | string | (optional) ClaimableBalanceID is a required parameter when claiming a Claimable Balance operation.

Reviewer 3: Same thing here, this needs to go in the actual response from the `POST /interactive` endpoints and referenced here.

Reviewer 3: Now that I'm thinking about this more, I'm realizing it's impossible for anchors to return a `claimable_balance_id` for a payment that hasn't been submitted to the network at this point. So I don't think we can have this response attribute at all.

The `claimable_balance_id` can be added to the [response](#deposit-and-withdraw-shared-responses)
This can be helpful for wallets to communicate to the user that their
transaction was submitted as a claimable balance and can be claimed at the relevant location (`claimable_balance_id`).

> In order to claim it the recipient must either get access to the account that the anchor created and or establish a trust-line (for the asset code of the claimableBalance) with their existing account.

Reviewer 3: `get access to the account that the anchor created` is a bit confusing. The wallet will already have access. The anchor creates the account but the wallet is the only one with the secret key.

I would remove this sentence, the flow is the same regardless of who created the account.

**Anchor Claimable Balance Recommended Options**

#### Claimable Balance Predicates

> [https://developers.stellar.org/docs/glossary/claimable-balance/#parameters](https://developers.stellar.org/docs/glossary/claimable-balance/#parameters)

Reviewer 3: Lets not do raw links in the SEP. We should also precede the link with context so they know what the link is before reading the URL or clicking it.

> We suggest if you decide to set predicates to claim the balance in 30 days. `BEFORE_RELATIVE_TIME(X)`

Reviewer 3: Did we decide 30 days was a good suggestion? Whats the rationale?

Author: We discussed that 30 days might be a good sweet spot to not keep balances piling up on the dex without them being claimed.

I'm still open to discussion on this one.

On one hand I think its totally fine to keep in unconditional and let the recipient claim on their own time whether it be a week a month or a year.
On the other hand, I think that if we help push recipients to claim their balance with a bit of haste it could lead them to do more activity (payments and whatnot) on the network. Instead of having balances eternally rot.

> However if implemented, its up to the anchors to handle refunds (of off network funds) to the the waller/user.

Reviewer 3: I think we need to expand on this. _why_ does this introduce the need to handle refunds? Why would an anchor want to have an expiration? Why do we recommend an expiration? This SEP is the introduction to this functionality for most people reading this.

Author: Correct me if I'm off the mark on this theoretical flow if the anchor sets the 30 day predicate.

1. Alice deposits off chain funds (US dollar)
2. Alice doesn't have a trust-line established for (USD) such that the anchor makes a claimable balance for Alice to claim (30 days to claim)
3. Alice doesnt claim the balance before 30 days and the balance is locked.
4. The anchor should refund Alice of her off chain funds

Author: Honestly after writing that and demoing what that behavior would be I think its probably for the best if we ensure that they set the predicates as unconditional.

Reviewer 3: I don't think we should recommend going either way, but outline why an anchor would take one approach over the other. 

Anchors that really really care about UX will probably want to handle refunds so users don't think they stole funds only to realize it's on them to "claim" the balance. 

For other anchors, handling refunds won't be worth it because there is no real benefit to them if they refund. The have to do 2X the logistical effort for no profit (fees are refunded).

Author: @tomerweller 
Would you be opposed to strongly suggesting unconditional claiming conditions?

5. `Anchor` detects the trust line creation in the `Account`. If the asset is `AUTH_REQUIRED`, `Anchor` approves the new trust line.
6. `Anchor` proceeds with the deposit flow.

> ##### Claimable Balances

Reviewer 3: Can you start with stating that claimable balances are currently optional by highly recommended? Also, note that in the future, claimable balances will be required.

Eventually, the hope is that all anchors will support CB's, and therefore we no longer need to support the 'waiting for trust-line' flow.

Author: How does the description read now?

> [Claimable Balances](https://www.stellar.org/blog/protocol-14-improvements) enables payments to accounts that do not have a trust line for the asset being sent.

Reviewer 3: ```suggestion
[Claimable Balances](https://www.stellar.org/blog/protocol-14-improvements) enable payments to accounts that do not have a trust line for the asset being sent.
```

> This feature provides less friction for Stellar accounts who wish to make an offchain deposit to anchors without making a trust line before the deposit is complete.

Reviewer 3: This doesn't make sense. Stellar accounts can't make off-chain deposits.

Author: I'll change it to something more appropriate
```suggestion
This feature provides less friction for Stellar accounts who wish to make a deposit external assets with an anchor without making a trust line before the deposit is complete.
```

> Anchors will no longer have to wait during the time when the account doesn't have a trust line. Instead, anchors can make the payment using ClaimableBalance operation and the stellar account (wallet user) can collect the funds at their own convenience.

Reviewer 3: ```suggestion
Anchors will no longer have to wait until the user's Stellar account has trust line to the sending asset before sending funds. Instead, anchors can make the payment using a ClaimableBalance operation and the user's Stellar account can claim the funds at their own convenience.
```

> **NOTE**: The Stellar Account will still need to establish trust line to the asset before claiming the outstanding balance.

Reviewer 3: maybe this belongs in the `Wallet Feature: Claim Account Claimable Balance` section? It seems a bit early to say this here.

**Wallets**: To support claimable balances wallets must

> * Send `claimable_balance_supported=true` in their POST /transactions/deposit/interactive request.

Reviewer 3: I wouldn't specify the `true` value like you are right now. It looks like a URL-encoded parameter, but client will be making a post request. Instead I'd provide a JSON example:
```json
{
    "claimable_balance": true
}
```

* Periodically poll for account's available claimable balances.

> * Provide a UI that allows users to claim a claimable balance. (The wallet UI should allow users to trust the asset of the claimable balance in order to claim)

Reviewer 3: I would leave out the trust-line/TX-construction details until you get to the `Claim Account Claimable Balance` section
```suggestion
* Provide a UI that allows users to claim a claimable balance.
```

**Anchors**: To support claimable balances anchors must

> * Accept the `claimable_balance_supported` field in their `POST /transactions/deposit/interactive` request and submit the deposit transaction as a Claimable Balance operation to Stellar account that don't yet have a trust-line to the asset they have requested.

Reviewer 3: Making the singular v. plural consistent
```suggestion
* Accept the `claimable_balance_supported` field in `POST /transactions/deposit/interactive` requests and submit deposit transactions using Claimable Balance operations to Stellar accounts that don't yet have a trust-line to the asset requested.
```

> * Add `claimable_balance_id` to their [deposit transactions](####Fields-for-deposit-transactions)

Reviewer 3: ```suggestion
* Add `claimable_balance_id` to their [deposit transactions responses](####Fields-for-deposit-transactions)
```

> **Anchors and Wallets**: Both anchors and wallets must also support the aforementioned ["stellar account doen't have a trust line for that asset."](#stellar-account-doesnt-trust-asset).

Reviewer 3: ```suggestion
**Anchors and Wallets**: Both anchors and wallets must also support the aforementioned ["stellar account doen't have a trust line for that asset."](#stellar-account-doesnt-trust-asset) flow.
```

Wallets need to be interoperable with anchors that have not adopted the [claimable balance feature](#Claimable-Balances) and vice versa.

##### Wallet Claimable Balance Flow

Wallets will still follow the [typical deposit flow](#Basic-Wallet-Implementation).

> Wallets will provide `claimable_balance_supported=true` field and value in their POST parameters for `/transactions/deposit/interactive`.

Reviewer 3: Same here with the URL-coded example

> If the anchor we're depositing to **supports** claimable balances.

Reviewer 3: Wallets unfortunately won't know if claimable balances are supported unless they hard-code it per-anchor. So i think we need to rewrite this **supports**/**doesn't support** section from the perspective that the wallet doesn't know.

That would involve sending `claimable_balance_supported: true`, then polling the service for `pending_trust`. If the status goes to `completed` without ever moving to `pending_trust`, the wallet knows to check for claimable balances, and the `claimable_balance_id` should be populated.

Author: The way I intended to frame `If the anchor we're depositing to **supports** claimable balances.` and `If the anchor we're depositing to ** doesn't support** claimable balances.` is from the wallet's perspective they do not know if the anchor their issuing a deposit to is `claimable_balance_supported` or not, they're just shooting their shot. 
Let me explain using the demo wallet implementation as an example as far as I understand it.

1. Wallet issues a deposit for a user (no trustline to depositing asset) and sends the `claimable_balance_supported` POST request param.
2. The wallet listens for the anchor's callback for the transaction status.
3. Until the transaction is completed we communicate its pending.
From the Wallet logs
```
To collect the interactive information we launch the interactive URL in a frame or webview, and await payment details from a postMessage callback
Transaction status pending...
Transaction status complete
```
If the anchor supports claimable balances it will do all the things it needs to do to process the claimable balance transaction.
Then mark it as complete

If the anchor doesnt support claimable balances (lets say it ignores the param) 
The wallet will keep listening to the transaction status from the callback.
```
To collect the interactive information we launch the interactive URL in a frame or webview, and await payment details from a postMessage callback
Transaction status pending...
Transaction status pending...
Transaction status pending...
```
On the interactive window from `sep24/transaction/more_info` we'll see the status `waiting for trust-line to be established`
from there the wallet is busy waiting (like before) but this time its because it `pending_trust`  until the user makes that trustline.
I reproduced this by sending `claimable_balance_supported: false` from my wallet and letting polaris(`316-claimable-balance-attempt2`) do it's trustline flow.

This is the wallet's callback loop 
```
window.onmessage = ({ data: { transaction } }) => {
      if (transaction.status === 'completed') {
        this.updateAccount()
        this.logger.instruction('Transaction status complete')
        finish()
      } else {
        this.logger.instruction('Transaction status pending...')
        setTimeout(() => {
          const urlBuilder = new URL(transaction.more_info_url)
          urlBuilder.searchParams.set('callback', 'postMessage')

          popup.location.href = urlBuilder.toString()
        }, 1000)
      }
    }
```

1. When the transaction status becomes `pending_user_transfer_start` send the required payment as described in the interactive webapp callback or the `/transaction` endpoint.
1. The anchor will report transaction status as `pending_anchor` then make the payment as a claimable balance operation.
1. After transaction is submitted the anchor will report transaction status as `completed`

If the anchor we're depositing to **doesn't support** claimable balances.

1. Make a request to `/transactions/deposit/interactive` and provide a value for the `claimable_balance_supported=true` field.
1. When the transaction status becomes `pending_user_transfer_start` send the required payment as described in the interactive webapp callback or the `/transaction` endpoint.
1. The anchor will ignore the `claimable_balance_supported` field, discover the account does not have a trust line, and report transaction status as `pending_trust`.
1. At this point both the wallet and anchor are waiting on the trust line to be established in order to continue the deposit payment. This is described here [steps 4-6](#stellar-account-doesnt-trust-asset)

##### Wallet Feature: Claimable Balance Polling

> Claimable Balance Polling allows the wallet to update the claimant that their is a balance available to claim and they may need to establish a [trust line](#stellar-account-doesnt-trust-asset) in order to do so.

Reviewer 3: ```suggestion
Polling allows the wallet to detect claimable balances and notify their users that their is a balance available to claim. See the [Claim Account Claimable Balance](#Wallet Feature: Claim Account Claimable Balance) section to learn more about how to claim a balance.
```

There are a few options as to how a wallet can poll and surface claimable balance information for their users.

1. The wallet can poll Horizon's [`/claimable_balances/`](https://developers.stellar.org/api/resources/claimablebalances/list/) list endpoint to pull any available claimable balances for the claimant.

   * *Recommended request* `GET /claimable_balances?claimant={:claimant}&order={asc,desc}&limit={1-200}` or [See List Claimable Balances](https://developers.stellar.org/api/resources/claimablebalances/list/)

> 2. The wallet can poll Horizon's [`/claimable_balances/`](https://developers.stellar.org/api/resources/claimablebalances/single/) single endpoint to pull the submitted transactions' available claimable balance(s) for the claimant(user) using the claimable_balance_id.

Reviewer 3: I don't think this option will be valuable for wallets, because they won't know the balance ID until they stream the transaction or poll the anchor for the `claimable_balance_id`. In either scenario, they won't need to make an additional Horizon request to get info on the balance, they can just immediately submit the claim op.

Not to mention that decoding the result XDR and getting the balance ID is complex, as evident by how much you had to write in order to explain it. I vote to remove this method.

      2. then pull the `claimable_balance_id` from the `xdr`
      3. Use the `claimable_balance_id` to query Horizon's [`/claimable_balances/`](https://developers.stellar.org/api/resources/claimablebalances/single/) single endpoint to pull all the information needed for the claimable balance.
      4. Provide UI to showcase that their deposit was successful and that their's an available claimable balance as well as enable users to [claim a claimable balance.](#wallet-feature-claim-account-claimable-balance)

      * Example of how a wallet could decode the `results_xdr`

      ```js
      const StellarBase = require('stellar-base');

      const { xdr } = StellarBase;

      const b64 = "AAAAAAAAAGQAAAAAAAAAAQAAAAAAAAAOAAAAAAAAAAA9/w3pAlg0fEoZp+eC9g0slbYtvoySN3i5y+F4WJY8sgAAAAA=";
      const tr = xdr.TransactionResult.fromXDR(b64, 'base64');

      const operationIndex = 0; // if you have multiple operations then you the index to correspond to he claimable balance operation
      console.log(tr.result().results()[operationIndex].value().createClaimableBalanceResult().balanceId().toXDR('hex'));
      ```

> ##### Wallet Feature: Claim Account Claimable Balance

Reviewer 3: I think it would be valuable to outline the operations needed right here. You can keep the docs reference if they want to check it out, but ideally they don't have to leave the SEP to know what the TX should look like. See SEP-10's description of challenge transactions for inspiration.

Author: I think I prefer to link both the documentation and showcase example code for example.
Similar to [SEP 20](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0020.md#example-code-for-linking)
Thoughts?

```suggestion
##### Wallet Feature: Claim Account Claimable Balance

It's recommended that Wallets provides a UI to enable accounts to claim their balances [polled](#wallet-feature-claimable-balance-polling)

[Example on how to claim a balance](https://developers.stellar.org/docs/glossary/claimable-balance/#claim-claimable-balance)
```javascript
const transaction = new TransactionBuilder(account, {
      fee: 100,
      networkPassphrase: this.network_passphrase,
    })
      .addOperation(
        Operation.claimClaimableBalance({
          balanceId,
        })
      )
      .setTimeout(0)
      .build()
    transaction.sign(keypair)
    this.logger.request(
      'Submitting claimClaimableBalance transaction',
      transaction
    )
    const result = await this.server.submitTransaction(transaction)
```
```

Reviewer 3: That works, but note your current code is missing a change trust op. In order to accept a claimable balance the user needs to have a trustline to the asset. The tricky piece is that the change trust op only needs to be included once per asset. You should explain this nuance in this section before showing the code.

It's recommended that Wallets have an feature that allows the account to claim the balances that were polled from Horizon.
[Example on how to claim a balance](https://developers.stellar.org/docs/glossary/claimable-balance/#claim-claimable-balance)

##### Anchor Claimable Balance Flow

1. Wallets make a request to `/transactions/deposit/interactive` and provide a value for the `claimable_balance_supported=true` field.

> 1. Anchors mark the transaction as [`claimable_balance_supported=true`](#deposit)

Reviewer 3: URL-encoding thing

> 1. Anchors discover that the stellar account that their sending the asset to doesnt have a trust line.

Reviewer 3: ```suggestion
1. Anchors detect that the user's Stellar account doesn't have a trust-line.
```

> 1. The anchor should then report transaction status as `pending_anchor` and make the payment as a claimable balance operation.

Reviewer 3: ```suggestion
1. The anchor should then report transaction status as `pending_anchor` and submit a Stellar transaction containing a create claimable balance operation
```

1. From there the anchor will continue the [deposit flow](#basic-anchor-implementation)

> ##### Anchor Feature: Extend Single Historical Transaction endpoint For Claimable Balances

Reviewer 3: They need to extend both `/transaction` and `/transactions`

5. `Anchor` detects the trust line creation in the `Account`. If the asset is `AUTH_REQUIRED`, `Anchor` approves the new trust line.
6. `Anchor` proceeds with the deposit flow.

##### Claimable Balances

> [Claimable Balances](https://www.stellar.org/blog/protocol-14-improvements) allows payments to accounts that do not have a trust-line for the asset sent.

Reviewer 3: ```suggestion
[Claimable Balances](https://www.stellar.org/blog/protocol-14-improvements) enable payments to accounts that do not have a trust-line for the asset being sent.
```

> **Wallets**: It's mandatory for wallets to support Claimable Balances and should poll for their users' received Claimable Balances so that users can "claim" them.

Reviewer 3: ```suggestion
**Wallets**: It's mandatory for wallets to support Claimable Balances. The methods for supporting Claimable Balances are outlined below.

> **Anchors**: It's optional for anchors to use submit Claimable Balances as deposit transactions to Stellar accounts that don't yet have a trust-line to the asset they have requested.

Reviewer: ```suggestion
**Anchors**: It's optional for anchors to submit Claimable Balances as deposit transactions to Stellar accounts that don't yet have a trust-line to the asset they have requested.
```

**Wallet Claimable Balance Flow**

Wallets will still follow the [typical deposit flow](#Basic-Wallet-Implementation).
The POST request however will extend to include a `claimable_balance_supported` parameter

> **Wallet Claimable Balance Polling**

Reviewer 3: Leigh also commented on this section, but the way I see it there are two methods for Wallets to detect claimable balances:

1. Checking Horizon
2. Checking the anchor's `GET /transaction(s)` endpoint

With the Horizon option, they could poll on demand or stream transactions to the users Stellar account.

Author: Yeah thats a cleaner way to split the options.

> Claimable Balance Polling allows the wallet to update the claimant that their is a balance available to claim and they may need to establish[trust-line](#stellar-account-doesnt-trust-asset) in order to do so.

Reviewer: Wallets should probably poll claimable balances as a general feature, but for SEP-24 I think the flow as an isolated experience shouldn't require it. The anchor should provide a claimable balance ID and the wallet should try and claim only that ID. Otherwise the UI/UX could become quite confusing. If for example a claimable balance comes in unrelated to the anchor at the same time for a different amount, that would be very confusing for the user and lead to a support ticket being opened about the other amount. From a security point-of-view we should also avoid arbitrary transactions impacting a well defined user flow as well.

Author: >Wallets should probably poll claimable balances as a general feature, but for SEP-24 I think the flow as an isolated experience shouldn't require it. 

I agree that is should be a general feature however it should be a required feature for wallets who intend to support claimable balances as a fallback to communicate with Wallet users that they have balances available.
It's currently implemented [here](https://github.com/stellar/stellar-demo-wallet/pull/49). 
I want to provide wallets as little friction as possible when interacting with anchors who may be still in the process of supporting claimable balances.

> The anchor should provide a claimable balance ID and the wallet should try and claim only that ID. 

We discussed this and we would like to have this feature when anchor adoption of claimable balances are relatively high.
Then we can require anchors send claimable balance IDs in their deposit response that Wallets can anticipate.
So we're attempting to strategize our phases for wallets and anchors not to overwhelm them with too many changes immediately.

> There are a few ways a wallet can support .

Reviewer: >There are a few ways a wallet can support **?????**.

1. The wallet can poll Horizon's [`/claimable_balances/`](https://developers.stellar.org/api/resources/claimablebalances/list/) list endpoint to pull any available claimable balances for the claimant.

   * *Recommended request* `GET /claimable_balances?claimant={:claimant}&order={asc,desc}&limit={1-200}` or [See List Claimable Balances](https://developers.stellar.org/api/resources/claimablebalances/list/)

> 1. Wallets could decode the `results_xdr` of the transaction submitted

Reviewer 3: I think we should leave this option out. Its complicated and the `/claimable_balances` endpoint polling is easy.

Author: I'm more in favor of leaving it in order to cover our bases. 
It is kinda verbose but this was info I had to figure out so wouldn't be better to share it in the SEP?
Or is there a better place to communicate this info?

Author: Im going to take it out and shorten it significantly.
The more I look at it the more I see how much of a mouthful it is.

   3. Use the `claimable_balance_id` to query Horizon's [`/claimable_balances/`](https://developers.stellar.org/api/resources/claimablebalances/single/) single endpoint to pull all the information needed for the claimable balance.

   * *Recommended request* `GET /claimable_balances/:claimable_balance_id`

   * Example of how a wallet could decode the `results_xdr`

    ```js
    const StellarBase = require('stellar-base');

    const { xdr } = StellarBase;

    const b64 = "AAAAAAAAAGQAAAAAAAAAAQAAAAAAAAAOAAAAAAAAAAA9/w3pAlg0fEoZp+eC9g0slbYtvoySN3i5y+F4WJY8sgAAAAA=";
    const tr = xdr.TransactionResult.fromXDR(b64, 'base64');

    const operationIndex = 0; // if you have multiple operations then you the index to correspond to he claimable balance operation
    console.log(tr.result().results()[operationIndex].value().createClaimableBalanceResult().balanceId().toXDR('hex'));
     ```

**Wallet Claim Account Claimable Balance**

It's recommended that Wallets have an feature that allows the account to claim the balances that were polled from Horizon.
[Example on how to claim a balance](https://developers.stellar.org/docs/glossary/claimable-balance/#claim-claimable-balance)

> **Anchor Claimable Balance Flow**

Reviewer 3: instead of **bolding** all these sections, I would use the markdown `####` strategy so readers can clearly see where the `Wallet` section ends and `Anchor` section starts.

1. If a user's Stellar account does not have a trust-line to the asset's issuer account, one must be added via the [Change Trust](https://developers.stellar.org/docs/start/list-of-operations/#claim-claimable-balance) operation.

> 2. The `Anchor` proceeds with the deposit flow via the [CAP-0023](../core/cap-0023.md) Create Claimable Balance operation.

Reviewer 3: Is this correct? Why would an anchor have to add a change_trust op to the transaction? In fact, I don't think this _can_ happen because the user's SK is required to change the user account's trustlines

The wallet must use the response fields in the following way to complete the withdrawal:

- `status`: `pending_user_transfer_start` means the user has given all necessary info to the anchor, and it is up to the wallet to send the actual stellar assets.

> - `withdraw_anchor_account`: send the withdrawal payment to this Stellar account.

Reviewer: Seems like there's a view formatting issues in this PR that could be fixed independent of a major change to the SEP. Could they be moved to a separate PR we can merge quickly?

`deposit_memo_type` | string | (optional) Type for the `deposit_memo`.
`from` | string | Sent from address, perhaps BTC, IBAN, or bank account.
`to` | string | Stellar address the deposited assets were sent to.

> `claimable_balance_id` | string | (optional) ClaimableBalanceID is a required parameter when claiming a Claimable Balance operation.

Reviewer: I found this description confusing. I'd avoid the term parameter since we use the term field in the title. The use of the term required is also confusing in the context of an optional field. And its not clear to me in this context what it means for it to be required when claiming the operation, since the operation isn't claimed. I think we should say what the field contains and when it is used.
```suggestion
`claimable_balance_id` | string | (optional) ID of the Claimable Balance containing the asset, included when an anchor is issuing the asset via a Claimable Balance.
```

`deposit_memo_type` | string | (optional) Type for the `deposit_memo`.
`from` | string | Sent from address, perhaps BTC, IBAN, or bank account.
`to` | string | Stellar address the deposited assets were sent to.

> `claimable_balance_id` | string | (optional) ID of the Claimable Balance containing the asset, included when an anchor is issuing the asset via a Claimable Balance.

Reviewer 3: I'm not sure what that second part means
```suggestion
`claimable_balance_id` | string | (optional) ID of the Claimable Balance used to send the asset initially requested. 
```
